pkg/router/controller/status.go

```diff
@@ -0,0 +1,238 @@
+			return false
+		}
+	}
+	now := nowFn()
+	condition.LastTransitionTime = &now
+	ingress.Conditions = []routeapi.RouteIngressCondition{condition}
+	return true
+}
+
+func ingressConditionTouched(ingress *routeapi.RouteIngress) *unversioned.Time {
+	var lastTouch *unversioned.Time
+	for _, condition := range ingress.Conditions {
+		if t := condition.LastTransitionTime; t != nil {
+			switch {
+			case lastTouch == nil, t.After(lastTouch.Time):
+				lastTouch = t
+			}
+		}
+	}
+	return lastTouch
+}
+
+// recordIngressConditionFailure updates the matching ingress on the route (or adds a new one) with the specified
+// condition, returning true if the object was modified.
+func recordIngressConditionFailure(route *routeapi.Route, name string, condition routeapi.RouteIngressCondition) (*routeapi.RouteIngress, bool, *unversioned.Time) {
+	for i := range route.Status.Ingress {
+		existing := &route.Status.Ingress[i]
+		if existing.RouterName != name {
+			continue
+		}
+		lastTouch := ingressConditionTouched(existing)
+		return existing, setIngressCondition(existing, condition), lastTouch
+	}
+	route.Status.Ingress = append(route.Status.Ingress, routeapi.RouteIngress{RouterName: name})
+	ingress := &route.Status.Ingress[len(route.Status.Ingress)-1]
+	setIngressCondition(ingress, condition)
+	return ingress, true, nil
+}
+
+// hasIngressBeenTouched returns true if the route appears to have been touched since the last time
+func (a *StatusAdmitter) hasIngressBeenTouched(route *routeapi.Route, lastTouch *unversioned.Time) bool {
+	glog.V(4).Infof("has last touch %v for %s/%s", lastTouch, route.Namespace, route.Name)
+	if lastTouch.IsZero() {
+		return false
+	}
+	old, ok := a.expected.Get(route.UID)
+	if !ok || old.(time.Time).Equal(lastTouch.Time) {
+		return false
+	}
+	return true
+}
+
+// recordIngressTouch
+func (a *StatusAdmitter) recordIngressTouch(route *routeapi.Route, touch *unversioned.Time, err error) {
+	switch {
+	case err == nil:
+		if touch != nil {
+			a.expected.Add(route.UID, touch.Time)
+		}
+	case errors.IsConflict(err):
+		a.expected.Add(route.UID, time.Time{})
+	}
+}
+
+// admitRoute returns true if the route has already been accepted to this router, or
+// updates the route to contain an accepted condition. Returns an error if the route could
+// not be admitted.
+func (a *StatusAdmitter) admitRoute(oc client.RoutesNamespacer, route *routeapi.Route, name string) (bool, error) {
+	ingress, updated := findOrCreateIngress(route, name)
+	if !updated {
+		for i := range ingress.Conditions {
+			cond := &ingress.Conditions[i]
+			if cond.Type == routeapi.RouteAdmitted && cond.Status == kapi.ConditionTrue {
+				glog.V(4).Infof("admit: route already admitted")
+				return true, nil
+			}
+		}
+	}
+
+	if a.hasIngressBeenTouched(route, ingressConditionTouched(ingress)) {
+		glog.V(4).Infof("admit: observed a route update from someone else: route %s/%s has been updated to an inconsistent value, doing nothing", route.Namespace, route.Name)
```

Comment:
Two router instances, one configured to set route subdomain to "foo.bar" and one to "bar.foo".  If you don't guard, they will endlessly try to write conflicting updates.  During a rolling update where "foo.bar" is transitioning to "bar.foo", you'll also have that behavior.  We record and observe mutations that appear to be made by someone else and use that to backoff.